src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -436,40 +423,100 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
         }
 
         /// <summary>
-        ///  Returns a byte blob that matches serialized content of single array initializer.    
-        ///  returns -1 if the initializer is null or not an array of literals
+        /// Generates a byte blob that matches serialized content of an array.
         /// </summary>
-        private int TryGetRawDataForArrayInit(BoundArrayInitialization initializer, out ImmutableArray<byte> data)
+        /// <returns>True if the blob is generated successfully, false otherwise</returns>
+        private bool TryGetRawDataForArray(BoundArrayCreation initializer, out ImmutableArray<byte> data, out int elementCount)
         {
-            data = default;
+            elementCount = GetArrayElementCount(initializer);
 
-            if (initializer == null)
+            if (elementCount < 0)
+            {
+                data = default;
+                return false;
+            }
+
+            if (elementCount == 0)
+            {
+                data = ImmutableArray<byte>.Empty;
+                return true;
+            }
+
+            var arrayType = (ArrayTypeSymbol)initializer.Type;
+            var elementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf();
+
+            if (initializer.InitializerOpt != null)
+            {
+                return TryGetRawDataForArrayInit(initializer.InitializerOpt, elementType, out data);
+            }
+
+            var elementSize = elementType.SpecialType.SizeInBytes();
+            if (elementSize == 0)
+            {
+                data = default;
+                return false;
+            }
+
+            var arraySizeInBytes = elementCount * elementSize;
```

Comment:
Could we overflow here?

Reply:
Yep, I suppose so.
Even not considering the possible overflow, should we limit the maximum blob size? It may be not a good idea to allocate a single-used 100MB array as a static field, regardless it's declared with or without an initializer.

Reply:
Assuming this optimization increases assembly size equal to the size of the array initializer, we should come up with some modest limit beyond which we decline to make the optimization. Maybe 128 KB? Someone with more runtime/metadata expertise than me may be able to recommend a better limit.

If in the future someone is negatively impacted by the limit we choose, we can consider adjusting it.

Reply:
Fixed.